noxfile.py

```diff
@@ -0,0 +1,151 @@
+"""
+Nox configuration file
+See https://nox.thea.codes/en/stable/config.html
+"""
+
+import sys
+from pathlib import Path
+
+import nox
+import toml
+
+CURRENT_PYTHON = sys.executable or f"{sys.version_info.major}.{sys.version_info.minor}"
+
+CONFIG = toml.load("pyproject.toml")
+DEPENDENCIES = CONFIG["project"]["dependencies"]
+OPTIONAL_DEPENDENCIES = CONFIG["project"]["optional-dependencies"]
+NOX_DEPENDENCIES = ("nox", "toml")
+
+
+# Global options
+nox.options.stop_on_first_error = False
+nox.options.error_on_missing_interpreters = False
+
+# Require support for tags
+nox.needs_version = ">=2022.8.7"
+
+
+@nox.session(python=CURRENT_PYTHON, tags=("test", "all"))
+def test(session):
+    """Run tests"""
+    session.install(
+        *DEPENDENCIES, *OPTIONAL_DEPENDENCIES["azure"], *OPTIONAL_DEPENDENCIES["test"]
+    )
+    session.run("python", "-m", "unittest", "discover")
+
+
+@nox.session(python=CURRENT_PYTHON, tags=("all",))
+def coverage(session):
+    """Check test coverage"""
+    session.install(
+        *DEPENDENCIES,
+        *OPTIONAL_DEPENDENCIES["azure"],
+        *OPTIONAL_DEPENDENCIES["test"],
+        "coverage",
+    )
+
+    session.run("coverage", "erase")
+    session.run("coverage", "run", "-m", "lisa")
+    session.run("coverage", "run", "--append", "-m", "unittest", "discover")
+    session.run("coverage", "report")
+
+
+@nox.session(python=CURRENT_PYTHON, tags=("format", "all"))
+def black(session):
+    """Run black"""
+    session.install("black")
+    session.run("black", ".")
+
+
+@nox.session(python=CURRENT_PYTHON, tags=("all"))
+def docs(session):
+    """Build docs"""
+    session.install(
+        *DEPENDENCIES,
+        *OPTIONAL_DEPENDENCIES["docs"],
+        *OPTIONAL_DEPENDENCIES["azure"],
+        *OPTIONAL_DEPENDENCIES["aws"],
+    )
+
+    session.run("sphinx-build", "-b", "html", "docs", "docs/_build/html")
+
+
+@nox.session(python=CURRENT_PYTHON, tags=("format", "all"))
+def isort(session):
+    """Run isort"""
+    session.install("isort")
+    session.run("isort", ".")
+
+
+@nox.session(python=CURRENT_PYTHON, tags=("lint", "all"))
+def flake8(session):
+    """Run flake8"""
+    session.install(
+        "flake8 >= 4.0.1",
```

Comment:
When should the package be added here or pyproject?

Reply:
It really depends on reuse. You'll see in a later commit this is moved to pyproject.toml so they can be installed in the dev virtual environment. Generally, if it's only needed in one place it can go in the nox session, but if we want to reuse a group of dependencies, it should go in  pyproject.toml.